src/main/asciidoc/security-refresh/security.adoc

```diff
@@ -0,0 +1,773 @@
+
+</web-app>
+----
+
+This XML essentially says that to access any URL that starts with "/rest" requires the caller to have the role "user". Roles are opaque strings; merely identifiers. It's fully up to the application how broad or fine-grained they are.
+
+Note that in Jakarta EE, internally these XML constraints are transformed into `Permission` instances and made available via a specific type of the Permission Store that we explained above. Knowledge about this transformation is only needed for very advanced use cases.
+
+The observant reader may wonder if XML is really the only option here, given the strong feelings that exist in parts of the community around XML. The answer is yes and no. Jakarta EE does define the `@RolesAllowed` annotation that could be used to replace the XML shown above, but only the legacy Enterprise Beans has specified a behaviour for this when put on an Enterprise Bean. Jakarta REST has done no such thing, although the JWT API in MicroProfile has defined this for REST resources. In Jakarta EE however this remains a vendor specific extension.
+
+There are also a number of annotations and APIs in Jakarta EE to set these kinds of constraints for individual Servlets, but those won't help us much either here.
+
+==== Declare the authentication mechanism
+
+[source,java]
+----
+@ApplicationScoped
+@BasicAuthenticationMechanismDefinition(realmName = "basicAuth")
+@DeclareRoles({ "user", "caller" })
+@ApplicationPath("/rest")
+public class ApplicationConfig extends Application {
+
+}
+----
+
+To declare the usage of a specific authentication mechanism, Jakarta EE provides `[XYZ]MechanismDefinition` annotations. Such an annotation is picked up by the security system, and in response to it a CDI bean that implements the `HttpAuthenticationMechanism` is enabled for it.
+
+The annotation can be put on any bean, but in a REST application it fits particularly well on the `Application` subclass that also declares the path for REST resources.
+
+==== Define the identity store
+
+Finally, let's define a very simple identity store that the security system can use to validate provided credentials for a Basic authentication:
+
+[source,java]
+----
+@ApplicationScoped
+public class TestIdentityStore implements IdentityStore {
+
+    public CredentialValidationResult validate(UsernamePasswordCredential usernamePasswordCredential) {
+        if (usernamePasswordCredential.compareTo("john", "secret1")) {
+            return new CredentialValidationResult("john", Set.of("user", "caller"));
+        }
+
+        return INVALID_RESULT;
+    }
+
+}
+----
+
+This identity store only "contains" the single identity (user) "john", with password "secret1" and roles "user" and "caller". Defining this kind of identity store is often the simplest way to get started. Note that Jakarta Security does not define a simple identity store out of the box, because there are questions whether that would promote security best practices.
```

Comment:
> However, do you mean to say that it might promote bad security practices?''

It can be both I guess. The intend is to say that everything in Jakarta Security promotes best practices, and people are not really sure if a simple identity store fits in with those best practices.

> I also wonder if this shouldn't be a callout as well.

Yes, good candidate for a callout too.

Reply:
Ok. Can you maybe reword this a bit, then?